src/ppx/grfx/grfx_swapchain.cpp

```diff
@@ -402,5 +573,93 @@ Result Swapchain::PresentHeadless(uint32_t imageIndex, uint32_t waitSemaphoreCou
+    grfx::ImagePtr         image         = mDeviceTarget.colorImages[imageIndex];
+    grfx::RenderPassPtr    renderPass    = mDeviceTarget.loadRenderPasses[imageIndex];
+
+    commandBuffer->TransitionImageLayout(image, PPX_ALL_SUBRESOURCES, grfx::RESOURCE_STATE_PRESENT, grfx::RESOURCE_STATE_RENDER_TARGET);
+    {
+        grfx::RenderPassBeginInfo beginInfo = {};
+        beginInfo.pRenderPass               = renderPass;
+        beginInfo.renderArea                = renderPass->GetRenderArea();
+        beginInfo.RTVClearCount             = 1;
+        beginInfo.RTVClearValues[0]         = {{0, 0, 0, 0}};
+        beginInfo.DSVClearValue             = {1.0f, 0xFF};
+
+        commandBuffer->BeginRenderPass(&beginInfo);
+        f(commandBuffer);
+        commandBuffer->EndRenderPass();
+    }
+    commandBuffer->TransitionImageLayout(image, PPX_ALL_SUBRESOURCES, grfx::RESOURCE_STATE_RENDER_TARGET, grfx::RESOURCE_STATE_PRESENT);
+}
+
+void Swapchain::RecordPreamble(uint32_t imageIndex)
+{
+    if (IsHeadless()) {
+        return;
+    }
+    if (mIsRecording[imageIndex]) {
+        return;
+    }
+    grfx::CommandBufferPtr commandBuffer = mCommandBuffers[imageIndex];
+    commandBuffer->Begin();
+    mIsRecording[imageIndex] = true;
+
+    if (IsIndirect()) {
+        grfx::ImagePtr      srcImage = mIndirectTarget.colorImages[imageIndex];
+        grfx::ImagePtr      dstImage = mDeviceTarget.colorImages[imageIndex];
+        grfx::RenderPassPtr dstClear = mDeviceTarget.clearRenderPasses[imageIndex];
+        commandBuffer->TransitionImageLayout(dstImage, PPX_ALL_SUBRESOURCES, grfx::RESOURCE_STATE_PRESENT, grfx::RESOURCE_STATE_RENDER_TARGET);
+        {
+            // Clear screen.
+            grfx::RenderPassPtr renderPass = dstClear;
+
+            grfx::RenderPassBeginInfo beginInfo = {};
+            beginInfo.pRenderPass               = renderPass;
+            beginInfo.renderArea                = renderPass->GetRenderArea();
+            beginInfo.RTVClearCount             = 1;
+            beginInfo.RTVClearValues[0]         = {{0.5, 0.5, 0.5, 0}};
+            beginInfo.DSVClearValue             = {1.0f, 0xFF};
+
+            commandBuffer->BeginRenderPass(&beginInfo);
+            commandBuffer->EndRenderPass();
+        }
+        commandBuffer->TransitionImageLayout(dstImage, PPX_ALL_SUBRESOURCES, grfx::RESOURCE_STATE_RENDER_TARGET, grfx::RESOURCE_STATE_COPY_DST);
+        {
+            grfx::ImageToImageCopyInfo imcopy = {};
+
+            imcopy.extent.x = std::min(mCreateInfo.width, mIndirectTarget.width);
+            imcopy.extent.y = std::min(mCreateInfo.height, mIndirectTarget.height);
+
+            // Copy rendered image.
+            // Note(tianc): this should be a image blit instead of copy.
+            commandBuffer->TransitionImageLayout(srcImage, PPX_ALL_SUBRESOURCES, grfx::RESOURCE_STATE_PRESENT, grfx::RESOURCE_STATE_COPY_SRC);
+            commandBuffer->CopyImageToImage(&imcopy, srcImage, dstImage);
```

Comment:
`vkCmdCopyImage` does not do scaling, but that's common one between vk and dx. `vkCmdBlitImage ` does not have a dx12 equivalent, so it will require more changes. This at least makes rendering at 4k possible on smaller screen without using headless, albeit cropped .

Reply:
can you create a bug and put some comments with the link to the bug? So if we change the knob to adjust the resolution that are not matching the FB resolution, visually it would look wrong? if the srcImg has a higher res than the dstImg, would the copy cause any mem corruption?